### ISSUE TITLE
Undefined variables and dicts

### ISSUE THREAD

**Issue by jbscare** [ansible/ansible#27268]
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

Not sure -- something that handles dereferencing variables, presumably?

##### ANSIBLE VERSION

```
ansible 2.4.0 (devel 12766571e9) last updated 2017/07/25 08:21:14 (GMT -400)
  config file = /ansible/dom/ansible.cfg
  configured module search path = [u'/home/jsmift/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/jsmift/src/ansible/lib/ansible
  executable location = /home/jsmift/src/ansible/bin/ansible
  python version = 2.7.12 (default, Sep  1 2016, 22:14:00) [GCC 4.8.3 20140911 (Red Hat 4.8.3-9)]
```

##### CONFIGURATION

None that seems related.

##### OS / ENVIRONMENT

Amazon Linux AMI release 2016.09

##### SUMMARY

If you define a variable that refers to an element in a dict, and that element's parent element isn't defined, you get an error when you try to dereference the variable.

##### STEPS TO REPRODUCE

This playbook:

```yaml
- hosts: localhost
  vars:
    mydict: { a: apple, b: banana }
    myfruit: "{{ mydict[myletter] }}"
  tasks:
    - debug: msg="My fruit is {{ myfruit }}"
      when: myfruit is defined
```

Run the playbook with '--extra-vars myletter=a' and then with '--extra-vars myletter=c'.

##### EXPECTED RESULTS

When run with '--extra-vars myletter=a', we expect a debug task to run and print the message "My fruit is apple".

When run with '--extra-vars myletter=c', we expect the debug task to be skipped, due to the 'when' statement.

##### ACTUAL RESULTS

In Ansible 2.2.1.0, we get the expected results.

In Ansible 2.3.0.0 (and devel, the version mentioned above), we get the expected results with '--extra-vars myletter=a', but with '--extra-vars myletter=c' we get

```
TASK [debug] *******************************************************************
fatal: [localhost]: FAILED! => {"failed": true, "msg": "The conditional check 'myfruit is defined' failed. The error was: error while evaluating conditional (myfruit is defined): {{ mydict[myletter] }}: 'dict object' has no attribute u'c'\n\nThe error appears to have been in '/home/jsmift/ansible/dom/playbooks/foo.yml': line 6, column 7, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n  tasks:\n    - debug: msg=\"My fruit is {{ myfruit }}\"\n      ^ here\nWe could be wrong, but this one looks like it might be an issue with\nmissing quotes.  Always quote template expression brackets when they\nstart a value. For instance:\n\n    with_items:\n      - {{ foo }}\n\nShould be written as:\n\n    with_items:\n      - \"{{ foo }}\"\n"}
```

**Comment by jbscare** [comment]
Another, simpler reproducer:

```yaml
- hosts: localhost 
  tasks: 
    - debug: msg=foo 
      when: foo is defined 
    - debug: msg=foo 
      when: foo['bar'] is defined 
```

This also works in Ansible 2.2.1.0, and not in anything later than that. (Both examples don't work in 2.2.2.0, if that helps any.)

**Comment by s-hertel** [comment]
Sorry for mistakenly thinking this was fixed and removing the fix in my previous PR for it. Verifying this problem still exists on devel. I've made the fix #27268 and added an integration test for it.

**Comment by s-hertel** [comment]
Sorry for mistakenly thinking this was fixed and removing the patch in my previous PR for it. Verifying this problem still exists on devel. I've made the fix #27298 and added an integration test.

**Comment by duck-rh** [comment]
Maybe I don't recall properly but I don't remember having this problem in the 2.4 series. In 2.5 the regression is present.

This problem also affects the behavior of `|default()`.

**Comment by BarbzYHOOL** [comment]
yes, it's a big problem I had to find stupid workarounds before I realized it was an ansible bug

**Comment by jmnavarrol** [comment]
It seems it also fails on jinja-based evalutations.  Something like this also fails when some intermediate key is absent:
param: "{{ dict is defined and dict['key'] is defined and dict['key']['subkey'] is defined | ternary(dict['key']['subkey'], 'other value') }}"

...or its relative:
param: "{{ dict is defined and 'key' in dict and 'subkey' in dict['key']| ternary(dict['key']['subkey'], 'other value')

**Comment by dsgnr** [comment]
Is this still an issue in 2.6.2?

	ansible 2.6.2
	  config file = /etc/ansible/ansible.cfg
	  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
	  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
	  executable location = /usr/bin/ansible
	  python version = 2.7.15rc1 (default, Apr 15 2018, 21:51:34) [GCC 7.3.0]

Here's part of my play:

	---
	  vars:
	    - exclude_db:
	      - "Database"
	      - "information_schema"
	      - "performance_schema"
	      - "mysql"

    - name: Backup databases
      mysql_db:
        state: dump
        name: "{{ item }}"
        target: "/backups/{{date}}/mysql/{{ item }}.sql"
        login_user: root
      with_items: "{{ dblist.stdout_lines | difference(exclude_db) }}"
      when: mysql_service_exists.stdout | regex_search('true')

Receiving:
`"msg": "Unable to look up a name or access an attribute in template string ({{ dblist.stdout_lines | difference(exclude_db) }}).\nMake sure your variable name does not contain invalid characters like '-': argument of type 'StrictUndefined' is not iterable"`

**Comment by jmnavarrol** [comment]
Just to clarify, my expectations are the following:

This task should print "not defined but valid syntax" in every case, no matter if any of my_dict, my_dict['key'] or my_dict['key']['subkey'] are **not** defined:
```
- name: This should work
  debug:
    msg: "This is a subkey value: {{ my_dict['key']['subkey'] | default('not defined but valid syntax') }}"
```

This should also print "not defined but valid syntax" in every case:
```
- name: This should also work
  debug:
    msg: 'not defined but valid syntax'
  when: not my_dict['key']['subkey']
  (or, alternatively)
  when: my_dict['key']['subkey'] is not defined
```

**Comment by gforster** [comment]
This is still an issue in 2.6.4

**Comment by michaalbert** [comment]
I'm also having this issue in 2.7.4.

**Comment by bcoca** [comment]
so the problem is that you are testing `myfruit` which is correctly identified as `defined` it is set to a template string, but this template string itself is using an undefined item, the `defined` test cannot account for that since it only looks at the variable itself.

**Comment by jbscare** [comment]
Does this simpler reproducer help?
```
- hosts: localhost 
  tasks: 
    - debug: msg=foo 
      when: foo is defined 
    - debug: msg=foo 
      when: foo['bar'] is defined 
```
That doesn't have anything to do with template strings or extra-vars or anything else.

This worked in Ansible 2.2.1.0, and not in anything later than that.

**Comment by mkrizek** [comment]
> Does this simpler reproducer help?
> 
> ```
> - hosts: localhost 
>   tasks: 
>     - debug: msg=foo 
>       when: foo is defined 
>     - debug: msg=foo 
>       when: foo['bar'] is defined 
> ```
> 
> That doesn't have anything to do with template strings or extra-vars or anything else.
> 
> This worked in Ansible 2.2.1.0, and not in anything later than that.

This has been implemented in Ansible `2.8`:
```yaml
- hosts: localhost
  gather_facts: no
  tasks:
    - debug:
        msg: foo
      when: foo is defined

    - debug:
        msg: foo
      when: foo['bar'] is defined
```

```
$ ansible-playbook issues/27268.yml 
PLAY [localhost] **********************************************************************************************

TASK [debug] **************************************************************************************************
skipping: [localhost]

TASK [debug] **************************************************************************************************
skipping: [localhost]

PLAY RECAP ****************************************************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=0    skipped=2    rescued=0    ignored=0
```

**Comment by sivel** [comment]
Closing per above.
